shpkpr/marathon/application.py

```diff
@@ -0,0 +1,74 @@
+# stdlib imports
+import json
+import os
+
+# third-party imports
+import jsonschema
+from cached_property import cached_property
+
+# local imports
+from shpkpr import exceptions
+
+
+def default_schema():
+    """Path to the default schema used to validate application definitions
+    before sending to marathon.
+    """
+    return os.path.join(os.path.dirname(os.path.realpath(__file__)), "schema", "app.json")
+
+
+class ValidationError(exceptions.ShpkprException):
+    exit_code = 2
+
+    def format_message(self):
+        return 'Unable to validate application: %s' % self.message
+
+
+class ApplicationError(exceptions.ShpkprException):
+    exit_code = 2
+
+
+class MarathonApplication(object):
```

Comment:
Yeah, in trying to keep the interface lightweight I may have gone too far.

I like the idea of having full objects with regular properties but the marshalling/unmarshalling code would get pretty verbose I think (maybe that's not a bad thing).

The [marathon](https://github.com/thefactory/marathon-python) library that this PR replaces actually has a really nice interface exactly as described, but its lack of any tests or updates for marathon >0.9.x scares me.

So I see two options here:

1. Switch back to the third-party dep but start contributing tests and docs upstream, implementing our own interface (Go style) to protect us from changes in the underlying implementation.
2. As you've suggested, treat the marathon app as a hash/dict and have an object that handles validation and anything else that's required.

`2` feels more lightweight and pythonic to me, treat the application data as a hash, passing that around and having a set of functions/objects that can process/validate it as needed.

`1` feels almost pointless as we'd end up having to reimplement a lot of `marathon`'s functionality and objects just to stop our abstraction leaking through.

Or maybe a third option, keep our thin wrapper around the marathon hash, but remove `__getitem__` and `__setitem__` which would prevent dictionary style access and add explicit properties to read/modify the state of the internal hash on the object?

e.g.

```python
class MarathonApp(object):

    def __init__(self, app_hash):
        self._app = app_hash

    @property
    def id(self):
        return self._app['id']

    @id.setter
    def id(self, value):
        self._app['id'] = value

>>> myapp = MarathonApp({'id': 'my-app'})
>>> myapp.id
"my-app"
```

We only use a few of the properties in `shpkpr` right now, so wouldn't be a huge task. The biggest issue I see with this approach is the nested objects (which would also be an issue if we implemented a full object with nested objects for all the subtypes (container/healthchecks/labels/etc).

I'm overthinking this, need to take a step back and consider the options properly before fixing.